### PR TITLE
Rework parent-join to not require access to DocumentMapper

### PR DESCRIPTION
**Pull request by romseygeek**
Parent joins work using a cluster of field mappers: the join field itself;
a set of subfields that allow multiple relationships between parents and
children to be defined; and a metadata field that acts to only allow a 
single join field per index to be defined.  The various queries and
aggregations that use this infrastructure retrieve the join field mapper
via a static method and then build themselves by pulling individual
relationship mappers from this main mapper.

Using mappers rather than MappedFieldTypes means that we need to
expose DocumentMapper at search time, which is something we are
trying to avoid.  This commit refactors things so that the join relations
are encapsulated in a Joiner object, which lives instead on the 
MappedFieldType associated with the metadata join field.  Rather than
using the ParentJoinFieldMapper and connected ParentIdFieldMappers,
we can now build queries and aggregations using this Joiner object,
retrieved via the QueryShardContext or AggregationContext using
a static helper method on Joiner itself.  In addition, the mappers are
refactored to parametrized form, and the Builders for the metadata and
relation mappers are removed as they are always constructed directly
by the join field mapper, and never via parsable user input.

### CODE REVIEW

**Comment by nik9000** [comment]
> Rework parent-join to not require access to DocumentMapper

hell yeah!

**Review by javanna — COMMENTED** [review]
This is a great change! Would it be possible to split it into smaller PRs? For instance one could be to move the mappers to parametrized field mapper, another one to replace the lookup mechanism so that looking up mappers is no longer required. Or are these two tied to each other?

**Comment by romseygeek** [comment]
I can probably split the parametrization out into a separate PR, although encapsulating the connection logic into the Relations class is a major part of it - I'll give it a go.

File: modules/parent-join/src/main/java/org/elasticsearch/join/mapper/Joiner.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+     * @return a filter for parent documents of a specific child type
+     */
+    public Query parentFilter(String childType) {
+        return new TermQuery(new Term(joinField, childrenToParents.get(childType)));
+    }
+
+    /**
+     * @return a filter for child documents of a specific parent type
+     */
+    public Query getChildrenFilter(String parentType) {
```
Comment (javanna) [inline modules/parent-join/src/main/java/org/elasticsearch/join/mapper/Joiner.java:123]: nit: is there a reason why this method starts with get while the two above don't?

File: modules/parent-join/src/main/java/org/elasticsearch/join/mapper/Joiner.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+     * Get the Joiner for this context, or {@code null} if none is configured
+     */
+    public static Joiner getJoiner(AggregationContext context) {
+        return getJoiner(context::isFieldMapped, context::getFieldType);
+    }
+
+    /**
+     * Get the Joiner for this context, or {@code null} if none is configured
+     */
+    public static Joiner getJoiner(Predicate<String> isMapped, Function<String, MappedFieldType> getFieldType) {
```
Comment (javanna) [inline modules/parent-join/src/main/java/org/elasticsearch/join/mapper/Joiner.java:63]: shall we make this package private?

File: modules/parent-join/src/main/java/org/elasticsearch/join/mapper/Joiner.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+    private final Map<String, Set<String>> parentsToChildren = new HashMap<>();
+    private final Map<String, String> childrenToParents = new HashMap<>();
+
+    private final String joinField;
+
+    /**
+     * Constructs a Joiner based on a join field and a set of relations
+     */
+    public Joiner(String joinField, List<Relations> relations) {
```
Comment (javanna) [inline modules/parent-join/src/main/java/org/elasticsearch/join/mapper/Joiner.java:86]: could this be package private?

File: modules/parent-join/src/main/java/org/elasticsearch/join/mapper/Joiner.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ * Utility class to help build join queries and aggregations, based on a join_field
+ */
+public class Joiner {
```
Comment (javanna) [inline modules/parent-join/src/main/java/org/elasticsearch/join/mapper/Joiner.java:44]: make this final?

File: modules/parent-join/src/main/java/org/elasticsearch/join/mapper/Relations.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        return Objects.equals(parent, relation.parent) &&
+            Objects.equals(children, relation.children);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(parent, children);
+    }
+
+    static List<Relations> parse(Object node) {
```
Comment (javanna) [inline modules/parent-join/src/main/java/org/elasticsearch/join/mapper/Relations.java:60]: parse looks unused?

File: server/src/main/java/org/elasticsearch/search/aggregations/support/AggregationContext.java [hunk 6]
```diff
@@ -187,6 +189,11 @@ public MappedFieldType getFieldType(String path) {
             return context.getFieldType(path);
         }
 
+        @Override
+        public boolean isFieldMapped(String field) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/AggregationContext.java:193]: you may also remove getMapper from this class , one less usage of getMapperService ;)

**Review by javanna — CHANGES_REQUESTED** [review]
I left some minor comments, looks great though, thanks for untangling this, I would not have known where to start ;)

**Review by javanna — APPROVED** [review]
LGTM

**Comment by romseygeek** [comment]
@elasticmachine update branch

**Pull request merged**